File: src/runtime/tests/capabilities-resolver-new-test.ts
/**
 * @license
 * Copyright 2020 Google LLC.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
import {assert} from '../../platform/chai-web.js';
import {Flags} from '../flags.js';
import {VolatileStorageKey} from '../storageNG/drivers/volatile.js';
import {RamDiskStorageKey} from '../storageNG/drivers/ramdisk.js';
import {DatabaseStorageKey, MemoryDatabaseStorageKey, PersistentDatabaseStorageKey, MemoryDatabaseStorageKeyFactory} from '../storageNG/database-storage-key.js';
import {StorageKey} from '../storageNG/storage-key.js';
import {ReferenceModeStorageKey} from '../storageNG/reference-mode-storage-key.js';
import {EntityType, ReferenceType} from '../type.js';
import {CapabilitiesResolver} from '../capabilities-resolver-new.js';
import {ArcId} from '../id.js';
import {Capabilities, Persistence, Ttl} from '../capabilities-new.js';
import {Schema} from '../schema.js';
import {assertThrowsAsync} from '../../testing/test-util.js';
import {DriverFactory} from '../storageNG/drivers/driver-factory.js';
import {Manifest} from '../manifest.js';

describe('Capabilities Resolver New', () => {
  after(() => DriverFactory.clearRegistrationsForTesting());

  type StorageKeyType = typeof VolatileStorageKey|typeof RamDiskStorageKey|typeof DatabaseStorageKey;
  function verifyReferenceModeStorageKey(key: StorageKey, expectedType: StorageKeyType) {
    assert.isTrue(key instanceof ReferenceModeStorageKey);
    const refKey = key as ReferenceModeStorageKey;
    assert.instanceOf(refKey.backingKey, expectedType,
        `Expected ${refKey.backingKey.constructor.name} to be instance of ${expectedType.name}`);
    assert.instanceOf(refKey.storageKey, expectedType,
        `Expected ${refKey.storageKey.constructor.name} to be instance of ${expectedType.name}`);
  }
  const entityType = new EntityType(new Schema(['Thing'], {result: 'Text'}));
  const referenceType = new ReferenceType(entityType);
  const handleId = 'h0';

  const unspecified = Capabilities.fromAnnotations();
  const inMemory = Capabilities.create([Persistence.inMemory()]);
  const inMemoryWithTtls = Capabilities.create([Persistence.inMemory(), Ttl.days(1)]);
  const onDisk = Capabilities.create([Persistence.onDisk()]);
  const onDiskWithTtl = Capabilities.create([Persistence.onDisk(), Ttl.minutes(30)]);

  it('fails creating keys with no factories', Flags.withDefaultReferenceMode(async () => {
    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test')});
    // Verify storage keys for none of the capabilities cannot be created.
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        unspecified, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        inMemory, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        inMemoryWithTtls, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        onDisk, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        onDiskWithTtl, entityType, handleId));
  }));

  it('creates volatile keys', Flags.withDefaultReferenceMode(async () => {
    // Register volatile storage key factory.
    // Verify only volatile (in-memory, no ttl) storage key can be created.
    VolatileStorageKey.register();
    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test')});
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        unspecified, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        inMemory, entityType, handleId), VolatileStorageKey);
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        inMemoryWithTtls, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        onDisk, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        onDiskWithTtl, entityType, handleId));
  }));

  it('creates keys with db only factories', Flags.withDefaultReferenceMode(async () => {
    DatabaseStorageKey.register();
    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test')});
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        unspecified, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        inMemory, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        inMemoryWithTtls, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        onDisk, entityType, handleId), PersistentDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        onDiskWithTtl, entityType, handleId), PersistentDatabaseStorageKey);
  }));

  it('creates keys with volatile and db factories', Flags.withDefaultReferenceMode(async () => {
    // Register database storage key factories. Verify all storage keys created as expected.
    VolatileStorageKey.register();
    DatabaseStorageKey.register();
    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test')});
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        unspecified, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        inMemory, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        inMemoryWithTtls, entityType, handleId), MemoryDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        onDisk, entityType, handleId), PersistentDatabaseStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
        onDiskWithTtl, entityType, handleId), PersistentDatabaseStorageKey);
  }));

  it('creates keys with custom factory', Flags.withDefaultReferenceMode(async () => {
    // Register volatile factory, pass in-memory database in constructor.
    VolatileStorageKey.register();
    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test'), factories: [new MemoryDatabaseStorageKeyFactory()]});
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
      unspecified, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
      inMemory, entityType, handleId), VolatileStorageKey);
    verifyReferenceModeStorageKey(await resolver.createStorageKey(
      inMemoryWithTtls, entityType, handleId), MemoryDatabaseStorageKey);
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        onDisk, entityType, handleId));
    await assertThrowsAsync(async () => await resolver.createStorageKey(
        onDiskWithTtl, entityType, handleId));
  }));

  it('creates keys for recipe with volatile and db factories', Flags.withDefaultReferenceMode(async () => {
    VolatileStorageKey.register();
    DatabaseStorageKey.register();
    const manifestStr = `
        recipe
          h0: create
          h1: create @ttl('2h')
          h2: create @persistent
          h3: create @persistent @ttl('30m')
          h4: create @queryable @ttl('30m')
    `;
    const recipe = (await Manifest.parse(manifestStr)).recipes[0];

    const resolver = new CapabilitiesResolver({arcId: ArcId.newForTest('test')});
    const verifyHandleKey = async (handle, expectedStorageKeyType) => {
        const fromOldCapabilities = Capabilities.fromOldCapabilities(handle.capabilities, handle.ttl);
        verifyReferenceModeStorageKey(await resolver.createStorageKey(
            fromOldCapabilities, entityType, handleId), expectedStorageKeyType);
        const capabilities = Capabilities.fromAnnotations(handle.annotations);
        verifyReferenceModeStorageKey(await resolver.createStorageKey(
            capabilities, entityType, handleId), expectedStorageKeyType);
    };
    await verifyHandleKey(recipe.handles[0], VolatileStorageKey);
    await verifyHandleKey(recipe.handles[1], MemoryDatabaseStorageKey);
    await verifyHandleKey(recipe.handles[2], PersistentDatabaseStorageKey);
    await verifyHandleKey(recipe.handles[3], PersistentDatabaseStorageKey);
    await verifyHandleKey(recipe.handles[4], MemoryDatabaseStorageKey);
  }));
});

File: javatests/arcs/core/storage/CapabilitiesResolverNewTest.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage

import arcs.core.common.ArcId
import arcs.core.data.CapabilitiesNew
import arcs.core.data.CapabilityNew.Persistence
import arcs.core.data.CapabilityNew.Ttl
import arcs.core.data.CapabilityNew.Shareable
import arcs.core.data.EntityType
import arcs.core.data.FieldType
import arcs.core.data.ReferenceType
import arcs.core.data.Schema
import arcs.core.data.SchemaFields
import arcs.core.data.SchemaName
import arcs.core.storage.CapabilitiesResolverNew.Options
import arcs.core.storage.api.DriverAndKeyConfigurator
import arcs.core.storage.keys.DatabaseStorageKey
import arcs.core.storage.keys.RamDiskStorageKey
import arcs.core.storage.keys.VolatileStorageKey
import arcs.core.storage.referencemode.ReferenceModeStorageKey
import com.google.common.truth.Truth.assertThat
import kotlin.test.assertFailsWith
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

/** Tests for [CapabilitiesResolverNew]. */
@RunWith(JUnit4::class)
class CapabilitiesResolverNewTest {
    private val entityType = EntityType(
        Schema(
            setOf(SchemaName("Thing")),
            SchemaFields(mapOf("name" to FieldType.Text), emptyMap()),
            "42"
        )
    )
    private val thingReferenceType = ReferenceType(entityType)
    private val handleId = "h0"

    private val unspecified = CapabilitiesNew()
    private val inMemory = CapabilitiesNew(listOf(Persistence.IN_MEMORY))
    private val inMemoryWithTtls = CapabilitiesNew(listOf(Persistence.IN_MEMORY, Ttl.Days(1)))
    private val onDisk = CapabilitiesNew(listOf(Persistence.ON_DISK))
    private val onDiskWithTtl = CapabilitiesNew(listOf(Persistence.ON_DISK, Ttl.Days(1)))
  
    @After
    fun tearDown() {
        CapabilitiesResolver.reset()
        CapabilitiesResolver.defaultCreators.clear()
        CapabilitiesResolverNew.reset()
    }

    private inline fun <reified T : StorageKey> verifyStorageKey(key: StorageKey) {
        assertThat(key).isInstanceOf(ReferenceModeStorageKey::class.java)
        assertThat((key as ReferenceModeStorageKey).backingKey is T).isTrue()
        assertThat(key.storageKey is T).isTrue()
    }

    @Test
    fun capabilitiesResolver_failsNonUniqueRegistration() {
        VolatileStorageKey.registerKeyCreator()
        assertFailsWith<IllegalArgumentException> {
            VolatileStorageKey.registerKeyCreator()
        }
    }

    @Test
    fun capabilitiesResolver_createStorageKey_failsUnsupported() {
        val resolver = CapabilitiesResolverNew(Options(ArcId.newForTest("test")))
        // Verify storage keys for none of the capabilities cannot be created.
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(unspecified, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(inMemory, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(inMemoryWithTtls, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(onDisk, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(onDiskWithTtl, entityType, handleId)
        }
    }

    @Test
    fun capabilitiesResolver_createsVolatileKeys() {
        // Register volatile storage key factory.
        VolatileStorageKey.registerKeyCreator()
        val resolver = CapabilitiesResolverNew(Options(ArcId.newForTest("test")))
        // Verify only volatile (in-memory, no ttl) storage key can be created.
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(unspecified, entityType, handleId)
        )
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(inMemory, entityType, handleId)
        )
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(inMemoryWithTtls, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(onDisk, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(onDiskWithTtl, entityType, handleId)
        }
    }

    @Test
    fun capabilitiesResolver_createsDatabaseKeys() {
        DatabaseStorageKey.registerKeyCreator()
        val resolver = CapabilitiesResolverNew(Options(ArcId.newForTest("test")))
        verifyStorageKey<DatabaseStorageKey.Memory>(
            resolver.createStorageKey(unspecified, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Memory>(
            resolver.createStorageKey(inMemory, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Memory>(
            resolver.createStorageKey(inMemoryWithTtls, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Persistent>(
            resolver.createStorageKey(onDisk, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Persistent>(
            resolver.createStorageKey(onDiskWithTtl, entityType, handleId)
        )
    }

    @Test
    fun capabilitiesResolver_createsAllKeys() {
        DriverAndKeyConfigurator.configureKeyParsers()
        val resolver = CapabilitiesResolverNew(Options(ArcId.newForTest("test")))
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(unspecified, entityType, handleId)
        )
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(CapabilitiesNew(listOf(Shareable(false))), entityType, handleId)
        )
        verifyStorageKey<RamDiskStorageKey>(
            resolver.createStorageKey(CapabilitiesNew(listOf(Shareable(true))), entityType, handleId)
        )
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(inMemory, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Memory>(
            resolver.createStorageKey(inMemoryWithTtls, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Persistent>(
            resolver.createStorageKey(onDisk, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Persistent>(
            resolver.createStorageKey(onDiskWithTtl, entityType, handleId)
        )
    }

    @Test
    fun capabilitiesResolver_createWithCustomFactories() {
        VolatileStorageKey.registerKeyCreator()
        val resolver = CapabilitiesResolverNew(
            Options(ArcId.newForTest("test")),
            listOf(DatabaseStorageKey.Memory.Factory())
        )
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(unspecified, entityType, handleId)
        )
        verifyStorageKey<VolatileStorageKey>(
            resolver.createStorageKey(inMemory, entityType, handleId)
        )
        verifyStorageKey<DatabaseStorageKey.Memory>(
            resolver.createStorageKey(inMemoryWithTtls, entityType, handleId)
        )
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(onDisk, entityType, handleId)
        }
        assertFailsWith<IllegalStateException> {
            resolver.createStorageKey(onDiskWithTtl, entityType, handleId)
        }
    }
}
